### ISSUE TITLE
Add HOWTO: Gradient clipping

### ISSUE THREAD

**Issue by marcvanzee** [google/flax#859]
We currently don't have a canonical implementation in Flax. Some optimizers have it built in.

Possible implementation:

```
def clip_grad_norm(grad, max_norm):
  norm = jnp.linalg.norm(jax.tree_util.tree_leaves(jax.tree_map(jnp.linalg.norm, grad)))
  factor = jnp.minimum(max_norm, max_norm / (norm + 1e-6))
  return jax.tree_map((lambda x: x * factor), grad)
```

**Comment by VolodyaCO** [comment]
Any progress on this?

**Comment by marcvanzee** [comment]
No, but we happily accept PRs!
@VolodyaCO do you want to take this?

**Comment by EyalRozenberg1** [comment]
I think a more accurate form for _clipping gradients_ would be:
```
def clip_grad_norm(grad, max_norm):
    norm = jnp.linalg.norm(jax.tree_util.tree_leaves(jax.tree_map(jnp.linalg.norm, grad)))
    clip = lambda x: jnp.where(norm < max_norm, x, x * max_norm / (norm + 1e-6))
    return jax.tree_util.tree_map(clip, grad)
```

**Comment by jheek** [comment]
Shouldn't this be part of optax?

**Comment by EyalRozenberg1** [comment]
Yes. 
I actually just saw that there is an implementation there for what we discussed.
`optax.clip_by_global_norm()`

**Comment by marcvanzee** [comment]
Ah good point @jheek. Closing this issue then (but please re-open if you think I'm wrong).
